[improve][pip] PIP-369: Flag based selective unload on changing ns-isolation-policy

iosdev747 opened the pull request:
Discussion thread: #23094 

<!-- Details of when a PIP is required and how the PIP process work, please see: https://github.com/apache/pulsar/blob/master/pip/README.md -->

### Motivation

In Apache Pulsar 3.x, changing a namespace isolation policy results in unloading all namespace bundles that match the namespace's regular expression provided in the isolation policy.
This can be problematic for cases where the regex matches a large subset of namespaces, such as `tenant-x/.*`. 
One of such case is mentioned on this issue [#23092](https://github.com/apache/pulsar/issues/23092) where policy change resulted in 100+ namespace bundles to get unloaded.
And broker exhausted all the available connections due to too many unload calls happening at once resulting in 5xx response.
Other issues that happens with this approach are huge latency spikes as topics are unavailable until bundles are loaded back, increasing the pending produce calls.
The only benefit this approach serves is ensuring that all the namespaces matching the policy regex will come to correct broker group. 
But when namespace bundles are already on the correct broker group (according to the policy), unloading those namespaces doesn't serve any purpose.

This PIP aims to address the need to either prevent unnecessary unloading or provide a more granular approach to determine what should be unloaded.

Some of the cases covered by this PIP are discussed in [#23094](https://github.com/apache/pulsar/issues/23094) by @grssam.
> - unload nothing as part of the set policy call
> - unload every matching namespace as part of the policy set call
> - unload only the changed namespaces (newly added + removed)

### Does this pull request potentially affect one of the following parts:

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: <!-- ENTER URL HERE -->

<!--
After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.
-->

lhotari commented:
Well written and explained. Thanks.

In the initial review, I only have a few knits about improving some concepts such as the `unload` enum in the proposal.
I think that putting more effort in the naming of it and the different values could be useful. For example, `all` should be something like `MATCHING` and the name of the `enum` should be more descriptive.

grssam commented:
> Well written and explained. Thanks.
> 
> In the initial review, I only have a few knits about improving some concepts such as the `unload` enum in the proposal. I think that putting more effort in the naming of it and the different values could be useful. For example, `all` should be something like `MATCHING` and the name of the `enum` should be more descriptive.

Thanks for the feedback. We can certainly name it `all_matching` . Notice that I am still using snake case to be consistent with some other enums used in the same policy call.
Generally the description of a flag is meant to contain more context. It's very tricky to convey what this param would do without exploding the param length.. How about something like `unload-scope` does that make more sense at first glance?

lhotari commented:
> Thanks for the feedback. We can certainly name it `all_matching` . Notice that I am still using snake case to be consistent with some other enums used in the same policy call.
> Generally the description of a flag is meant to contain more context. It's very tricky to convey what this param would do without exploding the param length.. How about something like `unload-scope` does that make more sense at first glance?

+1. Sounds good! `UnloadScope` would be a great concise name for the enum. Snake case values are fine if that's already the convention.

RobertIndie commented on pip/pip-369.md:
````diff
@@ -0,0 +1,121 @@
+
+```java
+enum unload {
+    all,        // unloads everything, OLD ⋃ NEW
````
Are there any benefits to using this flag? If not, how about removing this option?

grssam replied:
@RobertIndie - Are you referring about just the "all" option of the "unload" flag?

RobertIndie replied:
No. I'm suggesting removing the `all` option like this:
```java
enum unload {
    changed,    // unload namespaces delta, (new ⋃ old) - (new ∩ old) 
    none,       // skip unloading anything, ϕ
};
```

grssam replied:
that would be backward incompatible as since pulsar 2.7, it is doing what the `all` is meant to do.
We can discuss removing it in 4.x where we change the default to `changed`

RobertIndie replied:
Yeah.  What's the difference between `all` and `changed`? It seems that `all` offers no clear benefits, right? Or may I miss something? IMO, to users, this won't change the expected behavior, so we might not need to support this option.

Besides, there is a discussion for the 4.0.0 release: https://lists.apache.org/thread/mbrxy6jnz1yj0pyhmlr39wy0w1dhqv7s. The next feature release may be 4.0.0 instead of 3.4.0.

grssam replied:
`all` - all may fix all namespaces without making any assumptions. using the `all` option means that it ensures both the older and newer namespaces are in the right locattion.
`changed` takes an assumption that the unchanged namespaces are already in the correct location so it only unloads the namespaces which are either newly added or removed now.

Based on how things are, users can take a call to have force place every namespace or not.

Regarding releasing this - We want this to be present in 3.x as well since this majorly is a degradation in 2.11/3.x from 2.7/2.9 where there _was_ an option to disable this "unload all namespace" behavior. I believe all production systems will benefit from this present in the current LTS version, rather than making it wait for october time frame for 4.0 release.

The implementation PR is ready as well.

RobertIndie replied:
Thank you for clarifying. I had misunderstood the terms "new" and "old" before. Now I get it.

>  I believe all production systems will benefit from this present in the current LTS version

So you mean this PIP will be cherry-picked to the LTS version of 3.0.x?  I think we should discuss this separately or state it in the proposal. As we only allow bug fixes and important improvements cherry-picked into the released LTS version. 

And finally, I'm +1 for this PIP.

grssam replied:
> Thank you for clarifying. I had misunderstood the terms "new" and "old" before. Now I get it.

There is a catch for 3.x vs 4.0 in the meaning of "all" mentioned in the notes, again for compatibility reasons.

> So you mean this PIP will be cherry-picked to the LTS version of 3.0.x? I think we should discuss this separately or state it in the proposal. As we only allow bug fixes and important improvements cherry-picked into the released LTS version.

That's the idea. We will add it explicitly in this PIP proposal itself. 

> And finally, I'm +1 for this PIP.
Thank you for reviewing!

iosdev747 commented:
Thank you everyone for giving the feedback :)
I have addressed the review comments. And adding explicitly that this should get cherry picked in `3.0.x` as it provides a way to address the issue mentioned in https://github.com/apache/pulsar/issues/23092.

For changing the definition of `all/all_matching` and changing the default option to `changed`, it can be taken in a separate PR that will be present in 4.0+ but not in 3.0.x

I'll start a voting thread.

lhotari commented:
minor detail: `unload_scope` (snake case) as the enum name is perhaps against typical conventions. I believe that most enums are CamelCase, so instead of `unload_scope`, the enum type name should be `UnloadScope`.
